### PR TITLE
Period-based data usage reporting howto

### PR DESCRIPTION
**Pull request by terryburton**

### CODE REVIEW

File: doc/antora/modules/howto/nav.adoc [hunk 1]
```diff
@@ -59,3 +59,4 @@
 **** xref:eap-ttls.adoc[EAP-TTLS]
 **** xref:eap-peap.adoc[EAP-PEAP]
 **** xref:eap-tls.adoc[EAP-TLS]
+** xref:data-usage-reporting/index.adoc[Data Usage Reporting]
```
Comment (arr2036) [inline doc/antora/modules/howto/nav.adoc]: I think the probably belongs under the SQL section

File: doc/antora/modules/howto/pages/data-usage-reporting/index.adoc [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+without impacting normal usage reporting.
+
+For reporting purposes, the adhoc periods within the `data_usage_by_period`
+table can be efficiently aggregated to provide data bucketed into the required
+reporting intervals.
+
+
+== Example setup
+
+Megacorp stores their RADIUS accounting data in a MySQL database using the
```
Comment (arr2036) [inline doc/antora/modules/howto/pages/data-usage-reporting/index.adoc]: example.org

File: doc/antora/modules/howto/pages/data-usage-reporting/index.adoc [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+    mysql radius < /etc/raddb/mods-config/sql/main/mysql/process-radacct.sql
+
+NOTE: The above command assumes that passwordless login has been configured via
+the user's `~/.my.cnf` file, or otherwise. This command and subsequent commands
+should be adjusted accordingly where this is not the case.
+
+Call the update process manually to create the initial accounting period for
+the historical data usage and start of the next period for future use:
+
+    $ echo “CALL fr_new_data_usage_period()” | mysql radius
```
Comment (arr2036) [inline doc/antora/modules/howto/pages/data-usage-reporting/index.adoc]: Needs a code block (shell), fix smart quotes, remove dollar (means no one can paste the example into a terminal)

File: doc/antora/modules/howto/pages/data-usage-reporting/index.adoc [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+
+After some time, once a few sessions have reported in, call the update process
+again manually to populate the current period:
+
+    $ echo “CALL fr_new_data_usage_period()” | mysql radius
```
Comment (arr2036) [inline doc/antora/modules/howto/pages/data-usage-reporting/index.adoc]: Same as above

File: doc/antora/modules/howto/pages/data-usage-reporting/index.adoc [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+
+Ensure that data is being correctly populated. The data within the
+`data_usage_by_period` table will look something like this (for a single user):
+
+    mysql> SELECT * FROM data_usage_by_period;
```
Comment (arr2036) [inline doc/antora/modules/howto/pages/data-usage-reporting/index.adoc]: Should use example block

File: doc/antora/modules/howto/pages/data-usage-reporting/index.adoc [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+present in the `radacct` table when the update process was initial ran. Such rows
+can be removed at any time if data for this historic period is not required.
+
+Rows where `period_end` is `NULL` represent the open-ended period for future
+usage.  The `acctinputoctets` and `acctoutputoctets` for these rows will be
+zero or negative. _These rows must not be removed, otherwise data for ongoing
+sessions will be accounted for more than once._
+
+For this example we stated that the required reporting interval is initially
+monthly, however it is easy to postulate a future requirement to increase the
```
Comment (arr2036) [inline doc/antora/modules/howto/pages/data-usage-reporting/index.adoc]: Postulate is pushing the envelope for those who do not speak angleterreish as a primary language.

File: doc/antora/modules/howto/pages/data-usage-reporting/index.adoc [hunk 7]
```diff
@@ -0,0 +1,222 @@
...
+CAUTION: If storage requirements and resources allow then the frequency with
+which the `radacct` table is processed could be increased to hourly or even
+more often. However database "partitioning" should be considered for
+performance reasons if this would result in an enormous number of rows in the
+`data_usage_by_period` table.
+
+Next, enable daily processing of the `radacct` table by adding something like
+the following into the database user’s crontab:
+
+    0 12    * * *    echo “CALL fr_new_data_usage_period()” | mysql radius
```
Comment (arr2036) [inline doc/antora/modules/howto/pages/data-usage-reporting/index.adoc]: Code fence (cron)

File: doc/antora/modules/howto/pages/data-usage-reporting/index.adoc [hunk 8]
```diff
@@ -0,0 +1,222 @@
...
+
+To avoid storing usage data indefinitely a cronjob should be configured to
+clear out old accounting periods, such as the following:
+
+    15 3    * * *    echo “DELETE FROM data_usage_by_period WHERE \
```
Comment (arr2036) [inline doc/antora/modules/howto/pages/data-usage-reporting/index.adoc]: Same as above

Comment (arr2036) [reply doc/antora/modules/howto/pages/data-usage-reporting/index.adoc]: Smart quotes

File: doc/antora/modules/howto/pages/data-usage-reporting/index.adoc (hunk 7 above)
Comment (arr2036) [reply doc/antora/modules/howto/pages/data-usage-reporting/index.adoc]: Smart quotes

File: doc/antora/modules/howto/pages/data-usage-reporting/index.adoc [hunk 9]
```diff
@@ -0,0 +1,222 @@
...
+                           period_end < DATE_SUB(CURDATE(), INTERVAL 700 days); \
+                          ” | mysql radius
+
+The `data_usage_by period` table will now be maintained automatically.
+
+To extract the aggregated data usage of a user bucketed in monthly periods the
+`data_usage_by_period` table might be queried as follows:
+
+    SELECT
```
Comment (arr2036) [inline doc/antora/modules/howto/pages/data-usage-reporting/index.adoc]: Code fence (SQL)

File: doc/antora/modules/howto/pages/data-usage-reporting/index.adoc [hunk 10]
```diff
@@ -0,0 +1,222 @@
...
+periods, e.g. daily, weekly or monthly.
+
+The standard FreeRADIUS schema for accounting data (the `radacct` table) is
+keyed by session identifier (`acctuniqueid`) with each row representing a
+single session, containing (amongst other session data) the session start time
+(`acctstarttime`), the time that the last interim update was received
+(`acctupdatetime`), the session stop time (`acctstoptime`; if it has finished),
+and the total inbound and outbound data usage received in the last
+interim-update or stop packet (`acctinputoctets` and `acctoutputoctets`,
+respectively).
```
Comment (arr2036) [inline doc/antora/modules/howto/pages/data-usage-reporting/index.adoc]: I got out of breath reading that :)

**Comment by terryburton** [comment]
@arr2036 Thanks for the quick review. Agreed on all counts. I'll push changes tomorrow.

File: doc/antora/modules/howto/pages/data-usage-reporting/index.adoc [hunk 11]
```diff
@@ -0,0 +1,222 @@
...
+
+Megacorp stores their RADIUS accounting data in a MySQL database using the
+standard FreeRADIUS schema. They want to provide reports of monthly data usage
+for given users.
+
+Firstly, implement the extended schema to create the `data_usage_by_period`
+table to hold the data usage amounts and the `fr_new_data_usage_period` stored
+procedure that updates this table:
+
+    mysql radius < /etc/raddb/mods-config/sql/main/mysql/process-radacct.sql
```
Comment (arr2036) [inline doc/antora/modules/howto/pages/data-usage-reporting/index.adoc]: Shell code fence

File: doc/antora/modules/howto/pages/data-usage-reporting/index.adoc [hunk 12]
```diff
@@ -0,0 +1,222 @@
...
+granularity of the reports to some as yet unknown interval using the same
+historic usage data. We may therefore decide to process the `radacct` data
+each day, just after midnight to create fine-grained accounting periods with
+intervals that begin at the start of the day. Doing this allows accurate usage
+data to be reported for any interval that is some multiple of a day, yet no
+more often.
+
+CAUTION: If storage requirements and resources allow then the frequency with
+which the `radacct` table is processed could be increased to hourly or even
+more often. However database "partitioning" should be considered for
```
Comment (arr2036) [inline doc/antora/modules/howto/pages/data-usage-reporting/index.adoc]: Pick a quoting style and stick with it ;)

**Comment by terryburton** [comment]
Thanks.

**Comment by arr2036** [comment]
No problem, thanks for the docs!

**Pull request merged**